File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.rb
# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Integer[]}
def intersect(nums1, nums2)
    nums1tally = nums1.tally
    nums2tally = nums2.tally
    result=[]
    (nums1 & nums2).each do |num|
        [nums1tally[num], nums2tally[num]].min.times {result << num}
    end
    result
end

File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # Calculate frequency counts using Counter
        nums1count = Counter(nums1)
        nums2count = Counter(nums2)
        
        # Find common elements using set intersection
        common_elements = set(nums1count.keys()) & set(nums2count.keys())
        
        # Initialize result list
        result = []
        
        # Iterate through common elements
        for num in common_elements:
            # Determine the minimum count of num in nums1 & nums2 using cached tally results
            min_count = min(nums1count[num], nums2count[num])
            
            # Add num to result min_count times
            result.extend([num] * min_count)
        
        return result